File: crackup/drivers/CrackupDriverFile.php
<?php
/**
 * Filesystem storage driver for Crackup.
 * 
 * @author Ryan Grove <ryan@wonko.com>
 * @package Crackup
 */
class CrackupDriverFile extends CrackupDriver {
  // -- Public Instance Methods ------------------------------------------------
  public function delete($url) {
    if (@unlink($url)) {
      return true;
    }
    
    throw new CrackupDriverDeleteException('Unable to delete remote file: '.
        $url);
  }
  
  public function get($url, $local_filename) {
    if (@copy($url, $local_filename)) {
      return true;
    }
    
    throw new CrackupDriverGetException('Unable to get remote file: '.$url);
  }
  
  public function put($local_filename, $url) {
    if (@copy($local_filename, $url)) {
      return true;
    }
    
    throw new CrackupDriverPutException('Unable to put remote file: '.$url);
  }
}
?>

File: crackup/drivers/CrackupDriverFile.rb
require 'fileutils'

module Crackup
  class CrackupDriverFile < CrackupDriver
    def delete(url)
      File.delete(get_path(url))
      return true
    end
    
    def get(url, local_filename)
      FileUtils::copy(get_path(url), local_filename)
      return true
    end
    
    def put(url, local_filename)
      FileUtils::copy(local_filename, get_path(url))
      return true
    end
    
    private
    
    def get_path(url)
      uri  = URI::parse(url)
      path = ''
      
      unless uri.host.nil?
        path = "#{uri.host}:"
      end
      
      return path += uri.path
      
      rescue URI::InvalidURIError => e
        abort "Invalid URL: #{url}"
    end
  end
end
